utils.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import os
+import ssl
+import time
+from collections import Counter, defaultdict
+from typing import NamedTuple
+
+import pandas as pd
+from dotenv import load_dotenv
+from slack_sdk import WebClient
+from slack_sdk.errors import SlackApiError
+from sqlalchemy import create_engine
+from sqlalchemy.engine import Engine
+
+logging.basicConfig(format="%(asctime)s [%(levelname)s]:%(message)s", level=logging.INFO, datefmt="%Y-%m-%d %H:%M:%S")
+
+
+def mysql_connection() -> Engine:
+    """Connect to MySQL. This involves loading environment variables from file"""
+    os.chdir(os.path.dirname(os.path.abspath(__file__)))
+    load_dotenv()
+    engine = create_engine(
+        f"mysql+mysqlconnector://{os.getenv('DATABASE_USER')}:{os.getenv('DATABASE_PASSWORD')}@{os.getenv('DATABASE_HOST')}:3306"
+    )
+    return engine
+
+
+def slack_client(token: str) -> WebClient:
+    """Instantiate Slack Web client"""
+    ssl_context = ssl.create_default_context()
+    ssl_context.check_hostname = False
+    ssl_context.verify_mode = ssl.CERT_NONE
+    return WebClient(token=token, ssl=ssl_context)
+
+
+def _check_for_new_results(row: NamedTuple, year: int, idx: int, df: pd.DataFrame, awarded: pd.DataFrame) -> pd.DataFrame:
+    """Check for new earned achievements in the data. By looking at the current
+    achievement number and comparing it against what we've already seen, determine
+    if there are new achievments to issue. If there are no new achievments,
+    continue to the next one."""
```

Comment:
It would help with readability if we could have docstrings describing these inputs, especially the dataframes and what they contain

Reply:
Sure. Ran out of steam at the time but I'm in the process of adding those now.